components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,326 @@
+            name: name.to_owned(),
+            file_lock: Mutex::new(None),
+        };
+
+        let file_dict = log_file.recovery()?;
+        log_file.write(file_dict)?;
+        Ok(log_file)
+    }
+
+    /// Rewrite the log file to reduce file size and reduce the time of next recovery.
+    pub fn write(&self, file_dict: FileDictionary) -> Result<()> {
+        let mut append_file = self.file_lock.lock().unwrap();
+        let origin_path = self.base.join(&self.name);
+        let mut tmp_path = origin_path.clone();
+        tmp_path.set_extension(format!("{}.{}", thread_rng().next_u64(), TMP_FILE_SUFFIX));
+        let mut tmp_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(&tmp_path)
+            .unwrap();
+
+        let file_dict_bytes = file_dict.write_to_bytes()?;
+        let header = Header::new(&file_dict_bytes);
+        tmp_file.write_all(&header.to_bytes())?;
+        tmp_file.write_all(&file_dict_bytes)?;
+        tmp_file.sync_all()?;
+
+        // Replace old file with the tmp file aomticlly.
+        rename(&tmp_path, &origin_path)?;
+        let base_dir = File::open(&self.base)?;
+        base_dir.sync_all()?;
+        let file = std::fs::OpenOptions::new().append(true).open(origin_path)?;
+        append_file.replace(file);
+
+        Ok(())
+    }
+
+    /// Recovery from the log file and return `FileDictionary`.
+    pub fn recovery(&self) -> Result<FileDictionary> {
+        let mut _lock = self.file_lock.lock().unwrap();
+        let mut f = OpenOptions::new()
+            .read(true)
+            .open(self.base.join(&self.name))?;
+        let mut file_dict = FileDictionary::default();
+        let mut buf = Vec::new();
+        f.read_to_end(&mut buf)?;
+        let mut remained = buf.as_slice();
+
+        // parse the file dictionary
+        let (_, content) = Header::parse(remained)?;
+        remained.consume(Header::SIZE + content.len());
+        file_dict.merge_from_bytes(content)?;
+
+        // parse the remained records
+        while !remained.is_empty() {
+            let (used_size, file_name, mode) = parse_next_record(remained)?;
+            remained.consume(used_size);
+            match mode {
+                LogRecord::INSERT(info) => {
+                    file_dict.files.insert(file_name, info);
+                }
+                LogRecord::REMOVE => {
+                    file_dict.files.remove(&file_name);
+                }
+            }
+        }
+
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&self, name: String, info: FileInfo) -> Result<()> {
+        let mut file = self.file_lock.lock().unwrap();
+        if let Some(file) = file.as_mut() {
+            let bytes = convert_record_to_bytes(&name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
```

Comment:
we need to sync the parent directory, since the file size changed.

or there's another way, but somewhat more complicated. We can preallocate the file (say, 1G) (using fallocate, available in `nix::fallocate`), fsync the parent directory. then updating the file without changing its size do not need to fsync the parent directory and probably save some latency. but with it we will have no idea what's the actual file size.

up to you to decide which way to go. I don't have preference.

Reply:
If we don't care about the file_size just care about actual content read from the log file. Do we not need to sync the parent directory?

Reply:
The size is critical to integration. But seems size of the file is also flushed to disk by `fsync` or `fdatasync` on the file, so it's unnecessary to sync the parent directory. PTAL at https://linux.die.net/man/2/fsync.

Reply:
thanks @hicqu. I confused between fsync and fdatasync. When appending to a file fsync the is enough. There's no need to fsync the parent directory. The fallocate trick is to enable using fdatasync instead of fsync. When creating new file we do need to fsync the parent. Sorry for the misleading comment.